rfcs/taskforce/0003-Integration-test-infrastructure-spec.md

```diff
@@ -0,0 +1,214 @@
+
+In the current testnet infrastructure, a network manager would generate the necessary terraform configuration files and deploy using terraform.
+To run it locally, using minikube perhaps, we would create another network manager that deploys a testnet locally.
+
+The network manager itself can be specified as:
+
+```ocaml
+
+module Node : Node_intf
+(* Deploy the network*)
+val deploy : Network_config.t -> Daemon_config.t -> Testnet.t Deferred.t
+
+(*Tear down the network*)
+val destroy : Testnet.t -> unit Deferred.t
+```
+
+```ocaml
+
+module type Node_intf : sig
+    type t
+
+    val node_id_json : t -> json
+
+    val start : t -> unit Deferred.t
+
+    val stop : t -> unit Deferred.t
+
+    ...
+end
+
+module type Testnet = sig
+
+    type t =
+        {
+          block_producers : Node.t list
+        ; snark_coordinators: Node.t list
+        ; archive_nodes: Node.t list
+        ; ... }
+end
+```
+
+A node will encapsulate all the information needed to connect to it and make requests such as sending payments, retrieving a node's status, and so on. These requests implemented by the network manager could be using graphql commands. Given our intention to make all the cli commands use graphql queries, cli commands that do not currently have a graphql query should have that implemented first.
+Also, care should be taken to not add any test-specific queries/mutations/subscriptions that would not be used by users.
+
+The interface should also implement functions to start and stop a node that can be triggered at any time by the test program. This is particularly needed for testing functionalities like bootstrap and catchup. The network infrastructure should facilitate stopping and restarting coda daemon containers (with or w/o persisting the config directory). Currently a pod or a container within a pod is restarted automatically every time the daemon terminates. This behavior should be changed to support restarting a node on request. The following are some of the options to enable restart on request:
+    1. Have another process as the root of the container that initiates or terminates coda daemon based on a request from the network manager. This requires implementing and maintaining another application.
+    2. Using kubernetes init containers to start/stop coda daemon containers. The communication between the network manager and the init containers will have to be facilitated by incorporating some sort of message queues in the deployment architecture. This would not require a custom application that needs to be maintained but adds more complexity to the deployment architecture.
```

Comment:
Not really sure I see how `init` containers could be used in this case (based on the [docs](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#differences-from-regular-containers) and specifically due to `...must run to completion before the Pod can be ready.`)

Can somewhat see how this could work but the idea of being triggered only on deployments and as initialization logic seems to contradict the spec of **"...to start and stop a node that can be triggered at any time by the test program"**.

Reply:
The idea was that the init container would wait for a message to start the daemon container. But like @nholland94 mentioned it'll get complicated when trying to distinguish between crashes and intentional termination